src/iocost/iocost.c

```diff
@@ -0,0 +1,299 @@
+                }
+
+        return 1;
+}
+
+enum HwDbParseState {
+        MODEL = 1,
+        QOS = 2,
+        SKIP = 3,
+        DONE = 4,
+};
+
+static int hwdb_query_for_path(const char *path, sd_hwdb **hwdb, char **modalias) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        const char *model_name;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = sd_hwdb_new(hwdb);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open hwdb: %m");
+
+        if ((sd_device_get_property_value(device, "ID_MODEL_FROM_DATABASE", &model_name) < 0 &&
+             sd_device_get_property_value(device, "ID_MODEL", &model_name) < 0)) {
+                log_info("Model name for device %s is unknown", path);
+                return 0;
+        }
+
+        if (r < 0)
+                return log_error_errno(r, "Model name for device %s is unknown", path);
+
+        asprintf(modalias, "block:devname:%s:name:%s", path, model_name);
+
+        return 0;
+}
+
+static char *name_from_key(const char *key) {
+        _cleanup_strv_free_ char **key_parts = NULL;
+
+        key_parts = strv_split(key, "_");
+        return ascii_strlower(strdup(key_parts[2]));
+}
+
+static int apply_solution_for_path(const char *path, const char *name_to_apply) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *modalias = NULL;
+        _cleanup_(sd_hwdb_unrefp) sd_hwdb *hwdb = NULL;
+        enum HwDbParseState state = MODEL;
+        _cleanup_free_ char *name = NULL;
+        _cleanup_free_ char *qos = NULL;
+        _cleanup_free_ char *model = NULL;
+        const char *key, *value;
+        dev_t dev;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = sd_device_get_devnum(device, &dev);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        r = hwdb_query_for_path(path, &hwdb, &modalias);
+        if (r < 0)
+                return r;
+
+        SD_HWDB_FOREACH_PROPERTY(hwdb, modalias, key, value) {
+                if (state == DONE)
+                        break;
+
+                switch (state) {
+                        case MODEL:
+                                name = name_from_key(key);
+
+                                /* Not the parameters we want to apply, skip the QOS line and look for the next. */
+                                if (strcmp(name, name_to_apply)) {
+                                        state = SKIP;
+                                        break;
+                                }
+
+                                asprintf(&model, "%u:%u model=linear ctrl=user %s", major(dev), minor(dev), value);
```

Comment:
Missing OOM check.

Also, please use `DEVNUM_FORMAT_STR` and `DEVNUM_FORMAT_VAL` macros.